File: app/src/main/java/me/pwcong/jpstart/component/fragment/MemoryFragment.java
package me.pwcong.jpstart.component.fragment;

import android.graphics.Color;
import android.graphics.Paint;
import android.graphics.drawable.BitmapDrawable;
import android.graphics.drawable.Drawable;
import android.os.Bundle;
import android.support.annotation.Nullable;
import android.view.View;
import android.widget.RelativeLayout;

import com.amulyakhare.textdrawable.TextDrawable;
import com.getbase.floatingactionbutton.FloatingActionButton;
import com.getbase.floatingactionbutton.FloatingActionsMenu;

import java.util.List;

import butterknife.BindView;
import me.pwcong.jpstart.R;
import me.pwcong.jpstart.adapter.MemorySwipeAdapter;
import me.pwcong.jpstart.conf.Constants;
import me.pwcong.jpstart.manager.SoundPoolManager;
import me.pwcong.jpstart.mvp.bean.JPItem;
import me.pwcong.jpstart.mvp.presenter.BasePresenter;
import me.pwcong.jpstart.mvp.presenter.MemoryFragmentPresenterImpl;
import me.pwcong.jpstart.mvp.view.BaseView;
import me.pwcong.jpstart.utils.ResourceUtils;
import me.pwcong.jpstart.widget.dialog.ImageDialog;
import me.pwcong.jpstart.widget.swipecardview.SwipeFlingAdapterView;

/**
 * Created by Pwcong on 2016/9/29.
 */

public class MemoryFragment extends BaseFragment implements BaseView.MemoryFragmentView{

    @BindView(R.id.layout_root)
    RelativeLayout mRootLayout;
    @BindView(R.id.swipe_view)
    SwipeFlingAdapterView mSwipeFlingAdapterView;
    @BindView(R.id.fab_menu)
    FloatingActionsMenu mFabMenu;

    MemorySwipeAdapter adapter;
    BasePresenter.MemoryFragmentPresenter presenter;

    int category = Constants.CATEGORY_QINGYIN;

    @Override
    protected int getViewId() {
        return R.layout.fragment_memory;
    }

    @Override
    protected void initVariable(@Nullable Bundle savedInstanceState) {

        presenter=new MemoryFragmentPresenterImpl(this);

        initSwipeFlingAdapterView();
        initFabMenu();

    }


    private void initSwipeFlingAdapterView(){

        mSwipeFlingAdapterView.setFlingListener(new SwipeFlingAdapterView.onFlingListener() {
            @Override
            public void removeFirstObjectInAdapter() {
                adapter.remove(0);
                if (adapter.isEmpty()){
                    presenter.loadMore(category);
                }
            }

            @Override
            public void onLeftCardExit(Object dataObject) {

            }

            @Override
            public void onRightCardExit(Object dataObject) {

            }

            @Override
            public void onAdapterAboutToEmpty(int itemsInAdapter) {

            }

            @Override
            public void onScroll(float progress, float scrollXProgress) {

            }
        });


    }

    private void initFabMenu(){


        TextDrawable textQing = TextDrawable.builder()
                .beginConfig()
                .fontSize((int) getContext().getResources().getDimension(R.dimen.memory_item_fab_text_size))
                .endConfig()
                .buildRound("清", getContext().getResources().getColor(R.color.green));


        FloatingActionButton fab_qingyin = new FloatingActionButton(getContext());
        fab_qingyin.setIconDrawable(textQing);
        fab_qingyin.setColorNormal(getContext().getResources().getColor(R.color.green));
        fab_qingyin.setColorPressed(getContext().getResources().getColor(R.color.window));
        fab_qingyin.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {

                presenter.setDate(Constants.CATEGORY_QINGYIN);
                category=Constants.CATEGORY_QINGYIN;
                hideFabMenu();

            }
        });

        TextDrawable textZhuo = TextDrawable.builder()
                .beginConfig()
                .fontSize((int) getContext().getResources().getDimension(R.dimen.memory_item_fab_text_size))
                .endConfig()
                .buildRound("浊", getContext().getResources().getColor(R.color.orange));

        FloatingActionButton fab_zhuoyin = new FloatingActionButton(getContext());
        fab_zhuoyin.setIconDrawable(textZhuo);
        fab_zhuoyin.setColorNormal(getContext().getResources().getColor(R.color.orange));
        fab_zhuoyin.setColorPressed(getContext().getResources().getColor(R.color.window));
        fab_zhuoyin.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                presenter.setDate(Constants.CATEGORY_ZHUOYIN);
                category=Constants.CATEGORY_ZHUOYIN;
                hideFabMenu();
            }
        });

        TextDrawable textAo = TextDrawable.builder()
                .beginConfig()
                .fontSize((int) getContext().getResources().getDimension(R.dimen.memory_item_fab_text_size))
                .endConfig()
                .buildRound("拗", getContext().getResources().getColor(R.color.blue));

        FloatingActionButton fab_aoyin = new FloatingActionButton(getContext());
        fab_aoyin.setIconDrawable(textAo);
        fab_aoyin.setColorNormal(getContext().getResources().getColor(R.color.blue));
        fab_aoyin.setColorPressed(getContext().getResources().getColor(R.color.window));
        fab_aoyin.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                presenter.setDate(Constants.CATEGORY_AOYIN);
                category=Constants.CATEGORY_AOYIN;
                hideFabMenu();
            }
        });

        mFabMenu.addButton(fab_qingyin);
        mFabMenu.addButton(fab_zhuoyin);
        mFabMenu.addButton(fab_aoyin);

    }


    @Override
    protected void doAction() {
        presenter.initMemoryFragment();
    }

    @Override
    public void setData(List<JPItem> data) {

        if(adapter==null){
            adapter=new MemorySwipeAdapter(data);

            adapter.setOnWriteButtonClickListener(new MemorySwipeAdapter.OnWriteButtonClickListener() {
                @Override
                public void onClick(JPItem item) {

                    new ImageDialog.Builder(getContext())
                            .setResId(R.raw.write_a)
                            .override((int)ResourceUtils.getDimension(getContext(),R.dimen.dialog_width),
                                    (int)ResourceUtils.getDimension(getContext(),R.dimen.dialog_height))
                            .create()
                            .show();

                }
            });

            adapter.setOnYinButtonClickListener(new MemorySwipeAdapter.OnYinButtonClickListener() {
                @Override
                public void onClick(JPItem item) {
                    SoundPoolManager.getInstance().play(item.getRome());
                }
            });




        }
        adapter.setList(data);
        mSwipeFlingAdapterView.setAdapter(adapter);

    }

    @Override
    public void showMsg(int msg) {
        showSnackBar(mRootLayout,msg);
    }

    @Override
    public void showMsg(String msg) {
        showSnackBar(mRootLayout,msg);
    }

    @Override
    public void hideFabMenu() {
        mFabMenu.collapse();
    }

}

File: app/src/main/java/com/github/pwcong/jpstart/view/fragment/MemoryFragment.kt
package com.github.pwcong.jpstart.ui.fragment

import android.os.Bundle
import android.view.LayoutInflater
import android.view.ViewGroup

import com.github.pwcong.floatingactionbutton.FloatingActionButton
import com.github.pwcong.floatingactionbutton.FloatingActionsMenu
import com.github.pwcong.jpstart.R
import com.github.pwcong.jpstart.adapter.MemorySwipeAdapter
import com.github.pwcong.jpstart.constants.Constants
import com.github.pwcong.jpstart.databinding.FragmentMemoryBinding
import com.github.pwcong.jpstart.manager.GifManager
import com.github.pwcong.jpstart.manager.SoundPoolManager
import com.github.pwcong.jpstart.mvp.bean.JPItem
import com.github.pwcong.jpstart.mvp.presenter.BasePresenter
import com.github.pwcong.jpstart.mvp.presenter.impl.MemoryFragmentPresenterImpl
import com.github.pwcong.jpstart.mvp.view.BaseView.MemoryFragmentView
import com.github.pwcong.jpstart.ui.component.dialog.ImageDialog
import com.github.pwcong.jpstart.utils.ResourceUtils
import com.github.pwcong.swipecards.SwipeFlingAdapterView
import com.github.pwcong.textdrawable.view.TextDrawable

class MemoryFragment : BaseFragment<FragmentMemoryBinding>(), MemoryFragmentView {

    private lateinit var mSwipeFlingAdapterView: SwipeFlingAdapterView

    private lateinit var mFabMenu: FloatingActionsMenu

    private var adapter: MemorySwipeAdapter? = null
    private lateinit var presenter: BasePresenter.MemoryFragmentPresenter

    private var category: Int = Constants.CATEGORY_QINGYIN

    override fun initViewBinding(
        inflater: LayoutInflater,
        container: ViewGroup?
    ): FragmentMemoryBinding {
        return FragmentMemoryBinding.inflate(inflater, container, false)
    }

    override fun initVariable(savedInstanceState: Bundle?) {
        mFabMenu = getViewBinding().fabMenu
        mSwipeFlingAdapterView = getViewBinding().swipeView
        presenter = MemoryFragmentPresenterImpl(this)

        initSwipeFlingAdapterView()
        initFabMenu()
    }

    private fun initSwipeFlingAdapterView() {
        mSwipeFlingAdapterView.setFlingListener(object : SwipeFlingAdapterView.onFlingListener {
            override fun removeFirstObjectInAdapter() {
                adapter?.remove(0)
                if (adapter?.isEmpty == true) {
                    presenter.loadMore(category)
                }
            }

            override fun onLeftCardExit(p0: Any?) {
            }

            override fun onRightCardExit(p0: Any?) {
            }

            override fun onAdapterAboutToEmpty(p0: Int) {
            }

            override fun onScroll(p0: Float) {
            }
        })
    }

    private fun initFabMenu() {
        val textQing = TextDrawable.builder()
            .beginConfig()
            ?.fontSize(
                requireContext().resources.getDimension(R.dimen.memory_item_fab_text_size).toInt()
            )
            ?.endConfig()
            ?.buildRound("清", requireContext().resources.getColor(R.color.transparent))

        val fabQingyin = FloatingActionButton(requireContext())
        fabQingyin.setIconDrawable(textQing!!)
        fabQingyin.setColorNormal(requireContext().resources.getColor(R.color.green))
        fabQingyin.setColorPressed(requireContext().resources.getColor(R.color.window))
        fabQingyin.setOnClickListener {
            presenter.setDate(Constants.CATEGORY_QINGYIN)
            category = Constants.CATEGORY_QINGYIN
            hideFabMenu()
        }

        val textZhuo = TextDrawable.builder()
            .beginConfig()
            ?.fontSize(
                requireContext().resources.getDimension(R.dimen.memory_item_fab_text_size).toInt()
            )
            ?.endConfig()
            ?.buildRound("浊", requireContext().resources.getColor(R.color.transparent))

        val fabZhuoyin = FloatingActionButton(requireContext())
        fabZhuoyin.setIconDrawable(textZhuo!!)
        fabZhuoyin.setColorNormal(requireContext().resources.getColor(R.color.orange))
        fabZhuoyin.setColorPressed(requireContext().resources.getColor(R.color.window))
        fabZhuoyin.setOnClickListener {
            presenter.setDate(Constants.CATEGORY_ZHUOYIN)
            category = Constants.CATEGORY_ZHUOYIN
            hideFabMenu()
        }

        val textAo = TextDrawable.builder()
            .beginConfig()
            ?.fontSize(
                requireContext().resources.getDimension(R.dimen.memory_item_fab_text_size).toInt()
            )
            ?.endConfig()
            ?.buildRound("拗", requireContext().resources.getColor(R.color.transparent))

        val fabAoyin = FloatingActionButton(requireContext())
        fabAoyin.setIconDrawable(textAo!!)
        fabAoyin.setColorNormal(requireContext().resources.getColor(R.color.blue))
        fabAoyin.setColorPressed(requireContext().resources.getColor(R.color.window))
        fabAoyin.setOnClickListener {
            presenter.setDate(Constants.CATEGORY_AOYIN)
            category = Constants.CATEGORY_AOYIN
            hideFabMenu()
        }

        mFabMenu.addButton(fabQingyin)
        mFabMenu.addButton(fabZhuoyin)
        mFabMenu.addButton(fabAoyin)
    }

    override fun doAction() {
        presenter.initMemoryFragment()
    }

    override fun setData(data: List<JPItem>) {
        if (adapter == null) {
            adapter = MemorySwipeAdapter(data)

            adapter!!.setOnWriteButtonClickListener(object :
                MemorySwipeAdapter.OnWriteButtonClickListener {
                override fun onClick(item: JPItem) {
                    val gif = GifManager.getInstance().getJPGif(item.rome)
                    if (gif != null) {
                        ImageDialog.Builder(context!!)
                            .setResId(gif.hiragana)
                            .override(
                                ResourceUtils.getDimension(context!!, R.dimen.dialog_width).toInt(),
                                ResourceUtils.getDimension(context!!, R.dimen.dialog_height).toInt()
                            )
                            .create()
                            .show()
                    }
                }
            })

            adapter!!.setOnYinButtonClickListener(object :
                MemorySwipeAdapter.OnYinButtonClickListener {
                override fun onClick(item: JPItem) {
                    SoundPoolManager.getInstance().play(item.rome)
                }
            })
        }
        adapter!!.setList(data)
        mSwipeFlingAdapterView.adapter = adapter
    }

    override fun showMsg(msg: Int) {
        showSnackBar(msg)
    }

    override fun showMsg(msg: String) {
        showSnackBar(msg)
    }

    override fun hideFabMenu() {
        mFabMenu.collapse()
    }
}
